### DISCUSSION TITLE
HasManyThrough, once ->get() ... loses types? (Laravel 11.15)

### DISCUSSION THREAD

**Discussion by EliW** [larastan/larastan#1991] [Q&A]
OK, going to try to boil this down to a straight-forward situation (even though it isn't).   But basically the problem I have is that in a specific situation, it appears that even with everything declared as generics everywhere ... once I call `->get()` on a hasManyThrough connection ... it 'forgets' the model types.

So here's the situation, roughly ...  We have a model PropertyManager ... which has many Properties, which can each have many Leases ...   So in PropertyManager model we have:
```php
    public function leases(): HasManyThrough
    {
        return $this->hasManyThrough(Lease::class, Property::class);
    }
```

Now, in the code, we have a section that looks like this:  (current() is a scope)
```php
$company = /*valid query returning a propertymanager*/
$leases = $company->leases()->current()->get();
foreach ($leases as $lease) {
    $this->line('Analyzing lease ' . $lease->id);
}
```

This code, under phpstan:5 (and I'll add, suddenly with Laravel 11.15, under 11.14 it didn't complain about this) ... is giving the following error:
```
  106    Access to an undefined property Illuminate\Database\Eloquent\Model::$id.                                                                                                  
```

So somehow it doesn't understand that what it has, is an actual 'Lease' vs a generic Model

But if I try to change this to the following (just removing the ->get(), which doesn't work:
```php
$company = /*valid query returning a propertymanager*/
$leases = $company->leases()->current();
foreach ($leases as $lease) {
    $this->line('Analyzing lease ' . $lease->id);
}
```

I get this phpstan error:
```
  105    Argument of an invalid type Illuminate\Database\Eloquent\Relations\HasManyThrough<App\Models\Lease> supplied for foreach, only iterables are supported.  
```

So basically, phpstan understands that $company->leases() is in fact a `HasManyThrough<Lease>` ... but for some reason when I issue the `->get()` ... it forgets that completely.

I similarly tried changing it to $company->leases()->current()->getQuery() to see what might happen ... and specifically got back a:  `Builder<Model>` .... versus a `Builder<Lease>` in phpstan-errors.

Any idea what's going on here?  (and also why it suddenly started with laravel 11.15?  I ran into these issues when testing the 11.15 shift)

**Comment by calebdw** [comment]
I'm working on a draft PR (#1990) to fix this

**Reply by EliW** [reply]
ahhh, is bug w/ Laravel 11.15 ... reading the threads, I guess they incorporated your changes into Laravel before you had larastan ready ;) 

Lemme know if I can help at all.

**Reply by EliW** [reply]
Hey @calebdw  ... bringing the conversation back over here to not impact the PR ticket ...   But I've switched to using your branch, and have attempted to clear up everything to get level:6 working again on 11.19 ... and while the $id issues above are fixed, I'm running into another odd issue, and not sure what the solution is ... And they happen to be related to polymorphic relationships.

So to start off, a quick explanation of the model(s) here ... basically we have an 'AbstractPolicy' model, which is polymorphic to 3 types of models:  Policy, ThirdPartyPolicy, or Enrollment.     So here's the two issues I'm having now:

### The Morph Itself:
So I was now getting a bunch of errors about how the polymorphic relation *(called `policyable`) was only returning 'Model', and so basically every subsequent call to a real value was erroring in phpstan because of being against 'Model' instead of a specific model.   After a bit of research I found that I'm now supposed to give the list of possible morphs, so I updated the actual relation to have this annotation:
```php
    /**
     * @return MorphTo<Policy|ThirdPartyPolicy|Enrollment,$this>
     */
    public function policyable(): MorphTo
    {
        return $this->morphTo();
    }
```

This fixed all the lower cases where I was calling $this->policyable->PROPERTY to work ...   But now I get this error:
```
------ ---------------------------------------------------------------------------------------------------------------------------------- 
  Line   app/Models/AbstractPolicy.php                                                                                                     
 ------ ---------------------------------------------------------------------------------------------------------------------------------- 
  41     Method App\Models\AbstractPolicy::policyable() should return                                                                      
         Illuminate\Database\Eloquent\Relations\MorphTo<App\Models\Enrollment|App\Models\Policy|App\Models\ThirdPartyPolicy,               
         $this(App\Models\AbstractPolicy)> but returns Illuminate\Database\Eloquent\Relations\MorphTo<Illuminate\Database\Eloquent\Model,  
         $this(App\Models\AbstractPolicy)>.                                                                                                
 ------ ---------------------------------------------------------------------------------------------------------------------------------- 
```

So while I've now properly declared what 3 types might be returned ... it's complaining that the 'actual' return value ... isn't one of those but just generically 'Model'?   How do I fix that?

Relatedly, I have some issues in ONE of the polymorph end models, specifically Policy ... I'm getting these errors:
```
 ------ --------------------------------------------------------------------------------------------------------------------------- 
  Line   app/Models/Policy.php                                                                                                      
 ------ --------------------------------------------------------------------------------------------------------------------------- 
  169    Method App\Models\Policy::tenants() should return Illuminate\Database\Eloquent\Relations\BelongsToMany<App\Models\Tenant,  
         App\Models\AbstractPolicy> but returns Illuminate\Database\Eloquent\Relations\BelongsToMany<App\Models\Tenant,             
         App\Models\AbstractPolicy|null>.                                                                                           
  177    Method App\Models\Policy::leases() should return Illuminate\Database\Eloquent\Relations\BelongsToMany<App\Models\Lease,    
         App\Models\AbstractPolicy> but returns Illuminate\Database\Eloquent\Relations\BelongsToMany<App\Models\Lease,              
         App\Models\AbstractPolicy|null>.                                                                                           
 ------ --------------------------------------------------------------------------------------------------------------------------- 
```

So basically it's claiming that there 'might not' be a an AbstractPolicy when these return?   However, if I update my typehint/annotation to match that exactly, it instead complains that having `|null` in there doesn't match the signature of `BelowsToMany` in the first place ... so it doesn't work either way.

Here's the definitions of both those methods:
```php
    /**
     * @return BelongsToMany<Tenant,AbstractPolicy>
     */
    public function tenants(): BelongsToMany
    {
        return $this->abstractPolicy->tenants();
    }

    /**
     * @return BelongsToMany<Lease,AbstractPolicy>
     */
    public function leases(): BelongsToMany
    {
        return $this->abstractPolicy->leases();
    }

    /**
     * Get the parent abstract insurance policy
     */
    public function abstractPolicy(): MorphOne
    {
        return $this->morphOne(AbstractPolicy::class, 'policyable');
    }
```

And here's what the abstractPolicy version of Tenants & Leases look like:
```php
    /**
     * Get all tenants associated with the policy
     * @return BelongsToMany<Tenant,$this>
     */
    public function tenants(): BelongsToMany
    {
        return $this
            ->belongsToMany(Tenant::class, 'tenant_policy', 'policy_id', 'tenant_id')
            ->using(TenantPolicy::class)
            ->withPivot('type')
            ->withTimestamps();
    }

    /**
     * Get any leases associated with the policy
     * @return BelongsToMany<Lease,$this>
     */
    public function leases(): BelongsToMany
    {
        return $this->belongsToMany(Lease::class, 'lease_policy', 'policy_id', 'lease_id');
    }
```

But here's the crazy part?    Other variations of this work ... We have a `nonPrimaryTenants` relationship as well that's just a filtered version.   And it's not giving the same error/warning even though it's set up identically?

In Policy.php
```php
    /**
     * @return BelongsToMany<Tenant,AbstractPolicy>
     */
    public function nonPrimaryTenants(): BelongsToMany
    {
        return $this->abstractPolicy->nonPrimaryTenants();
    }
```

In AbstractPolicy:
```php
    /**
     * @return BelongsToMany<Tenant,$this>
     */
    public function nonPrimaryTenants(): BelongsToMany
    {
        return $this->tenants()->wherePivot('type', '!=', PRIMARY);
    }
```

The only difference is that it's actually calling the original relationship then doing a wherePivot on it.   But it should be the exact same return type, and it doesn't complain there.

I hope all this makes sense ...

**Reply by calebdw** [reply]
@EliW, 

There's a lot to unpack here but I'll do my best.

## PHPStan inside vs outside methods
Before we can get into specifics, it's important to understand how PHPStan actually works.

When PHPStan is analyzing your code, it can't descend into functions/methods every time they're called to see what they return---this would be wildly inefficient. Thus when PHPStan is **outside** of a method, it uses the method signature to verify that the right parameter types are being passed to the method and determine what the return type is.

When PHPStan is **inside** a method, it verifies that you are using the parameter types appropriately and it validates that you are returning what you say the method returns in the signature.

## MorphTo
The MorphTo relation is inherently mysterious---there is no way to determine ahead of time what model(s) it will return. You will notice that there is no classes that are passed to the method like other relations. Therefore, the only thing that we can say with certainty is that the MorphTo will return a Model.

Adding a union like this will work for analysis when PHPStan is **outside** the method, but you will still get an error when **inside** the method because the MorphTo returns a Model:

```php
    /**
     * @return MorphTo<Policy|ThirdPartyPolicy|Enrollment,$this>
     */
    public function policyable(): MorphTo
    {
        return $this->morphTo();
    }
```

However, this is very brittle and will break if you ever add another morphable type. Your two options are to:
1. keep the union and just ignore the error
2. update the type to `MorphTo<Model,$this>` and then use the Model public api everywhere that the model is used

2) is the better choice as it's not brittle, but it does require you to use models differently than you might be used to---you will need to use the public api:
```php
// before
$model->id;
$model->foo;
// after
$model->getKey();
$model->getAttribute('foo');
```

or you will need to narrow the type at the call site:

```php
if (is_a($model, User::class)) {
    // code where $model is a User
} elseif (...) {
}
// or 
assert($model instanceof User, 'the model must be a User');
```



## AbstractPolicy

I can't be sure without seeing how your Policy is defined, but it looks like your Policy does not extend the AbstractPolicy so the relations are returning exactly what they should be given that they are defined on the AbstractPolicy

**Reply by calebdw** [reply]
Ah, I know what's wrong with your relations---you need to call the scopes on the **method** not the property:

```diff
    /**
-    * @return BelongsToMany<Tenant,AbstractPolicy>
+    * @return BelongsToMany<Tenant,$this>
     */
    public function tenants(): BelongsToMany
    {
-       return $this->abstractPolicy->tenants();
+       return $this->abstractPolicy()->tenants();
    }

    /**
-    * @return BelongsToMany<Lease,AbstractPolicy>
+    * @return BelongsToMany<Lease,$this>
     */
    public function leases(): BelongsToMany
    {
-       return $this->abstractPolicy->leases();
+       return $this->abstractPolicy()->leases();
    }
```

When called on the method, the scope is applied to the relation, when called on the property the scope is applied to the model

**Reply by EliW** [reply]
(Reading/responding to these in opposite order)

So in your second response you said I needed to call `abstractPolicy()` instead of just `abstract` ...

But that causes a different issue ... since that now, instead of it realizing that it's calling `AbstractPolicy::leases` ... makes it give this error:
```
Call to an undefined method Illuminate\Database\Eloquent\Relations\MorphOne<App\Models\AbstractPolicy,                     
         $this(App\Models\Policy)>::leases().        
```

**Reply by calebdw** [reply]
Can you show your Policy and AbstractPolicy definitions? I'm not sure why your Policy class has an abstractPolicy relation on it---typically your Policy class should extend your AbstractPolicy

**Reply by EliW** [reply]
Now responding to the first (MorphTo) post:

> However, this is very brittle and will break if you ever add another morphable type.

True, but honestly I'd want that?   Since in the end if we did add another morphable type, I want to force us to 'document' that via the annotations.

> keep the union and just ignore the error

Currently what I'm doing ...

> update the type to MorphTo<Model,$this> and then use the Model public api everywhere that the model is used

OK, I can see how this would work, but it definitely feels very 'ugh', since all the accessors that exist in the polymorph such as this: 

```php
    public function provider()
    {
        return $this->policyable->provider();
    }
    public function isValid()
    {
        return $this->policyable->isValid();
    }
```

Are all throwing errors that you can't call those against `Model` ... So like you said, we'd have to wrap all of these in semi-annoying "if policyable is one of these three classes, call this method".    

I get why it's needed, but in practice the better solution seems to be *(1) ... because it's at the higher level having us just say:  "policyable will always be one of these 3 classes" ... and that instantly fixes all dozen+ accessor methods that call individual instance methods on the polymorphs.

**Reply by EliW** [reply]
> Can you show your Policy and AbstractPolicy definitions? I'm not sure why your Policy class has an abstractPolicy relation on it---typically your Policy class should extend your AbstractPolicy

I mean, I potentially can, but might need to scrub them, and they are both LARGE ... so I was trying to just share the 'important bits'.

As for extending, let's take a step back and realize that we in some cases don't have things named 'the best' here and just working with what we have.  AbstractPolicy in this case is NOT an abstract class ...   But it's the 'polymorphic table'.     It should have been more accurately named 'PolymorphicPolicy'.

Or likely just 'Policy'.    Then the three 'sub types' that it could polymorph into be named:   'InternalPolicy, ThirdPartyPolicy, EnrollmentPolicy' ...    

But that just wasn't what was done, so don't read too much into the 'abstract' part of the name there.

**Reply by calebdw** [reply]
Instead of specifying specific models, the better pattern is to use an interface and intersection type that enforces the specific models to implement the common methods:

```php
    /**
     * @return MorphTo<Model&Policyable,$this>
     */
    public function policyable(): MorphTo
    {
        return $this->morphTo();
    }
```

Of course, you might need to still ignore the error about this returning a `MorphTo<Model>` (but you might not)
